Macros for outputs (OpenSim_DECLARE_OUTPUT())

chrisdembia opened the pull request:
I've created three macros for creating outputs:

- OpenSim_DECLARE_OUTPUT
- OpenSim_DECLARE_OUTPUT_FLEX
- OpenSim_DECLARE_OUTPUT_FOR_STATE_VARIABLE

They work just fine, and once this is merged, then other people can start using these macros in prep for the hackathon.  To be consistent, I tried to follow our conventions for properties pretty closely.

I also documented these macros so they appear nicely in doxygen, just like properties. You can view the doxygen here: http://myosin.sourceforge.net/806/classOpenSim_1_1Model.html

I didn't spend too much time on the documentation because it is likely to change when we add multi-outputs and create similar macros for inputs, connectors, etc.

Other than the FLEX macro above, I don't think these macros need to take comments, since member functions and state variables should be documented elsewhere.

Two tasks that might be good to do but that I'd prefer to leave to a subsequent PR:

1. Allow default value for the dependsOn stage through the macros; this requires overloaded/variadic macros and is somewhat complicated but is definitely do-able.
2. The member variable created by `constructOutput()` is an int whose value is always 0. This doesn't make too much sense, but it may also be OK. Luckily, the macro lets us change this in the future, and so I'd rather leave that for a separate PR. Also, we can deal with macro-generated accessor methods in a separate PR (probably after the hackathon).

Once this is merged, I'll move all existing `constructProperty()` calls to the macro.

Also, we can create new macros for list outputs once we figure out how to do those.

chrisdembia commented:
Here's what the doxygen documentation looks like for outputs:
![image](https://cloud.githubusercontent.com/assets/846001/13551782/e13806ca-e2fa-11e5-9736-05283471589c.png)

sherm1 commented:
:+1:

chrisdembia commented:
Here's a question: should the macro automatically fill in the `SimTK::Stage::` part of the stage? That way, people would invoke the macro as:

```cpp
OpenSim_DECLARE_OUTPUT(force, double, getForce, Dynamics);
```

aseth1 commented on OpenSim/Common/Component.cpp:
```diff
@@ -820,9 +820,9 @@ setDiscreteVariableValue(SimTK::State& s, const std::string& name, double value)
 }
 
-void Component::constructOutputForStateVariable(const std::string& name)
+int Component::constructOutputForStateVariable(const std::string& name)
```
Should we settle now to use `SimTK_DEFINE_UNIQUE_INDEX_TYPE()`? instead of an int?
With the potential of having many indices I like using the unique index to thwart accidentally mixing indices up.

aseth1 commented on OpenSim/Common/ComponentOutput.h:
```diff
@@ -164,6 +164,114 @@ class Output : public AbstractOutput {
+/// near where you declare @ref Property properties.
+/// @{
+/** Create an output for a member function of the same component. 
+ *  The following must be true about componentMemberFunction, the function
```
same -> this component?

aseth1 commented on OpenSim/Common/ComponentOutput.h:
```diff
@@ -164,6 +164,114 @@ class Output : public AbstractOutput {
+/// Use these macros at the top of your component class declaration,
+/// near where you declare @ref Property properties.
+/// @{
+/** Create an output for a member function of the same component. 
```
same -> this component.

chrisdembia commented:
I've addressed your feedback. I've created an OutputIndex class, but I still set its value to 0 for all outputs. After the hackathon, we should revisit what we do with this return value.

chrisdembia commented:
Updated to replace OutputIndex with bool.

aseth1 commented:
Don't forget to remove OpenSim_DECLARE_OUTPUT_FLEX in a subsequent PR,

aseth1 commented on OpenSim/Common/ComponentOutput.h:
```diff
@@ -164,6 +164,114 @@ class Output : public AbstractOutput {
+    /** @}                                                               */ \
+    /** @cond                                                            */ \
+    bool _unused_output_##oname { constructOutput<T>(#oname, &Self::func, ostage) }; \
+    /** @endcond                                                         */
```
minor but instead of `_unused_output...` you could rename it to `has_output_...` since it is a bool anyway. Might be less confusing than having a bool member named `_unused` and has a value of `true`.

chrisdembia commented:
Updated; tests pass.

aseth1 commented:
Looks great!

chrisdembia commented:
Thank you.
